### DISCUSSION TITLE
Cannot find module 'firebase-admin/app' from　""

### DISCUSSION THREAD

**Discussion by yun8boo** [firebase/firebase-admin-node#1866] [Q&A]
When using Node.js SDK Admin SDK v10 in Nest.js, the test gets angry that the module is not found

```
    Cannot find module 'firebase-admin/app' from 'libs/firebase-admin/firebase-admin.service.ts'

    Require stack:
      libs/firebase-admin/firebase-admin.service.ts
      libs/firebase-admin/auth/auth.module.ts
      libs/firebase-admin/auth/auth.service.spec.ts

      1 | import { Injectable, Logger } from '@nestjs/common';
      2 | import { ConfigService } from '@nestjs/config';
    > 3 | import { ServiceAccount, initializeApp, cert, getApps } from 'firebase-admin/app';
        | ^
      4 | import { getAuth } from 'firebase-admin/auth';
      5 |
      6 | @Injectable()

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (libs/firebase-admin/firebase-admin.service.ts:3:1)
```

**Comment by MILLERMARRU** [comment]
This error comes from Jest not knowing how to handle Firebase Admin SDK v10+'s modular package exports (`"exports"` field in `package.json`). Jest's default module resolver doesn't follow the `exports` map, so `firebase-admin/app` can't be found.

## Fix for Jest + NestJS (or any Node project)

Add `moduleNameMapper` to your Jest config to manually map the modular imports:

```js
// jest.config.js
module.exports = {
  // ... your existing config
  moduleNameMapper: {
    '^firebase-admin/app$': '<rootDir>/node_modules/firebase-admin/lib/app/index.js',
    '^firebase-admin/auth$': '<rootDir>/node_modules/firebase-admin/lib/auth/index.js',
    '^firebase-admin/firestore$': '<rootDir>/node_modules/firebase-admin/lib/firestore/index.js',
    '^firebase-admin/messaging$': '<rootDir>/node_modules/firebase-admin/lib/messaging/index.js',
    '^firebase-admin/storage$': '<rootDir>/node_modules/firebase-admin/lib/storage/index.js',
    // Add other subpackages as needed
  },
};
```

## Alternative: use `transformIgnorePatterns`

If you're using ESM transforms, add Firebase Admin to the exceptions list:

```js
// jest.config.js
module.exports = {
  transformIgnorePatterns: [
    'node_modules/(?!(firebase-admin)/)',
  ],
};
```

## For NestJS projects specifically

NestJS uses `jest.config.js` at the root and sometimes a separate one inside `apps/`. Make sure the fix is in the config that Jest actually loads for the failing test file.

Also verify your `tsconfig.json` has `"moduleResolution": "node16"` or `"bundler"` — older `"node"` resolution doesn't follow package exports either, which can cause the same error at compile time (not just test time).

## Root cause

Firebase Admin SDK v10 switched to modular imports (following the `firebase` JS SDK pattern). Before v10, you imported everything from `'firebase-admin'`. The new subpath imports (`firebase-admin/app`, `firebase-admin/auth`, etc.) rely on the `exports` field in `package.json`, which Jest < 28 ignores by default. Jest 28+ respects `exports` if you set `"resolver": "jest-resolve"` with `"customExportConditions": ["node", "require"]`.
